app/scripts/lib/rpc-method-middleware/handlers/add-ethereum-chain.js

```diff
@@ -38,287 +35,134 @@ async function addEthereumChainHandler(
-      }),
-    );
-  }
-
-  function isLocalhostOrHttps(urlString) {
-    const url = getValidUrl(urlString);
-
-    return (
-      url !== null &&
-      (url.hostname === 'localhost' ||
-        url.hostname === '127.0.0.1' ||
-        url.protocol === 'https:')
-    );
-  }
-
-  const firstValidRPCUrl = Array.isArray(rpcUrls)
-    ? rpcUrls.find((rpcUrl) => isLocalhostOrHttps(rpcUrl))
-    : null;
-
-  const firstValidBlockExplorerUrl =
-    blockExplorerUrls !== null && Array.isArray(blockExplorerUrls)
-      ? blockExplorerUrls.find((blockExplorerUrl) =>
-          isLocalhostOrHttps(blockExplorerUrl),
-        )
-      : null;
-
-  if (!firstValidRPCUrl) {
-    return end(
-      ethErrors.rpc.invalidParams({
-        message: `Expected an array with at least one valid string HTTPS url 'rpcUrls', Received:\n${rpcUrls}`,
-      }),
-    );
-  }
-
-  if (blockExplorerUrls !== null && !firstValidBlockExplorerUrl) {
-    return end(
-      ethErrors.rpc.invalidParams({
-        message: `Expected null or array with at least one valid string HTTPS URL 'blockExplorerUrl'. Received: ${blockExplorerUrls}`,
-      }),
-    );
-  }
-
-  const _chainId = typeof chainId === 'string' && chainId.toLowerCase();
-
-  if (!isPrefixedFormattedHexString(_chainId)) {
-    return end(
-      ethErrors.rpc.invalidParams({
-        message: `Expected 0x-prefixed, unpadded, non-zero hexadecimal string 'chainId'. Received:\n${chainId}`,
-      }),
-    );
-  }
+  const existingNetwork = findExistingNetwork(
+    chainId,
+    findNetworkConfigurationBy,
+  );
 
-  if (!isSafeChainId(parseInt(_chainId, 16))) {
+  if (
+    existingNetwork &&
+    existingNetwork.chainId === chainId &&
+    existingNetwork.ticker !== ticker
+  ) {
     return end(
       ethErrors.rpc.invalidParams({
-        message: `Invalid chain ID "${_chainId}": numerical value greater than max safe value. Received:\n${chainId}`,
+        message: `nativeCurrency.symbol does not match currency symbol for a network the user already has added with the same chainId. Received:\n${ticker}`,
       }),
     );
   }
 
-  const existingNetwork = findNetworkConfigurationBy({ chainId: _chainId });
-
-  // if the request is to add a network that is already added and configured
-  // with the same RPC gateway we shouldn't try to add it again.
-  if (existingNetwork && existingNetwork.rpcUrl === firstValidRPCUrl) {
-    // If the network already exists, the request is considered successful
-    res.result = null;
-
-    const currentChainId = getCurrentChainId();
-    const currentRpcUrl = getCurrentRpcUrl();
+  let networkConfigurationId;
+  let requestData;
+  let approvalFlowId;
 
-    // If the current chainId and rpcUrl matches that of the incoming request
-    // We don't need to proceed further.
-    if (currentChainId === _chainId && currentRpcUrl === firstValidRPCUrl) {
-      return end();
-    }
+  if (!existingNetwork || existingNetwork.rpcUrl !== firstValidRPCUrl) {
```

Comment:
This will also change after https://github.com/MetaMask/core/pull/4286. Network configurations will have multiple RPC endpoints and also a default RPC endpoint to use. I'm not sure exactly how this will look but you may be able to simplify this.